piv/key.go

```diff
@@ -194,23 +194,38 @@ func Verify(attestationCert, slotCert *x509.Certificate) (*Attestation, error) {
 }
 
 type verifier struct {
-	Root *x509.Certificate
+	Roots *x509.CertPool
 }
 
 func (v *verifier) Verify(attestationCert, slotCert *x509.Certificate) (*Attestation, error) {
-	root := v.Root
-	if root == nil {
-		ca, err := yubicoCA()
+	o := x509.VerifyOptions{KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageAny}}
+	o.Roots = v.Roots
+	if o.Roots == nil {
+		cas, err := yubicoCAs()
 		if err != nil {
-			return nil, fmt.Errorf("parsing yubico ca: %v", err)
+			return nil, fmt.Errorf("failed to load yubico CAs")
 		}
-		root = ca
+		o.Roots = cas
 	}
-	if err := verifySignature(root, attestationCert); err != nil {
-		return nil, fmt.Errorf("attestation certifcate not signed by : %v", err)
+
+	o.Intermediates = x509.NewCertPool()
+
+	// the attestation cert in some yubikey 4 does not encode X509v3 Basic Constraints.
+	// this isn't valid as per https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9
+	// (fourth paragraph) and thus makes x509.go validation fail. work around this by
+	// setting this constraint here.
+	//
+	// an example of such attestation cert can be found as a test vectors in key_test.go#TestVerify
```

Comment:
This comment would be better in the test file? E.g.

```
// Attestation chain from a YubiKey 4 that does not encode X.509 v3 Basic Constrains.
name: "InvalidChainBasicConstraints",
```

Reply:
Addressed, thanks